maggy/optimizer/bayes/base.py

```diff
@@ -0,0 +1,1016 @@
+                    )
+                    self._log("sampled randomly: {}".format(hparams))
+                else:
+                    if self.pruner and not self.interim_results:
+                        # sample from largest model available
+                        model_budget = max(self.models.keys())
+                    # sample from model with model budget
+                    self._log(
+                        "start sampling routine from model with budget {}".format(
+                            model_budget
+                        )
+                    )
+                    hparams = self.sampling_routine(model_budget)
+                    next_trial = self.create_trial(
+                        hparams=hparams,
+                        sample_type="model",
+                        run_budget=run_budget,
+                        model_budget=model_budget,
+                    )
+                    self._log(
+                        "sampled from model with budget {}: {}".format(
+                            model_budget, hparams
+                        )
+                    )
+
+            # check if Trial with same hparams has already been created
+            i = 0
+            while self.hparams_exist(trial=next_trial):
+                self._log("sample randomly to encourage exploration")
+                hparams = self.searchspace.get_random_parameter_values(1)[0]
+                next_trial = self.create_trial(
+                    hparams=hparams, sample_type="random_forced", run_budget=run_budget
+                )
+                i += 1
+                if i > 3:
+                    self._log(
+                        "not possible to sample new config. Stop Experiment (most/all configs have already been used)"
+                    )
+                    return None
+
+            # report new trial id to pruner
+            if self.pruner:
+                self.pruner.report_trial(
+                    original_trial_id=None, new_trial_id=next_trial.trial_id
+                )
+            self._log(
+                "start trial {}: {}, {} \n".format(
+                    next_trial.trial_id, next_trial.params, next_trial.info_dict
+                )
+            )
+            return next_trial
+
+        except BaseException:
+            self._log(traceback.format_exc())
+            self._close_log()
+
+            if self.pruner:
+                self.pruner._close_log()
+
+    def finalize_experiment(self, trials):
+        self._log("Experiment finished")
+        self._close_log()
+
+        if self.pruner:
+            self.pruner._close_log()
+
+        return
+
+    def init_model(self):
+        """initializes the surrogate model of the gaussian process
+
+        the model gets created with the right parameters, but is not fit with any data yet. the `base_model` will be
+        cloned in `update_model` and fit with observation data
+        """
+        raise NotImplementedError
```

Comment:
can we make this an abstractmethod so it's clear it's being implemented by the specific BO implementations?

Reply:
yes